File: Overoom.React/src/services/room-hub/room.hub.ts
import * as signalR from '@microsoft/signalr';
import { HubConnection } from '@microsoft/signalr';

import { RoomEventContainer } from './events/room-event.container.ts';

/** Класс для работы с SignalR хабом комнаты */
export class RoomHub {
  /** Массив обработчиков событий комнаты */
  private readonly handlers: ((event: RoomEventContainer) => void)[];

  /** URL endpoint SignalR хаба */
  private readonly url: string;

  /** Экземпляр подключения SignalR */
  private connection?: HubConnection;

  /** Фабрика для получения JWT токена аутентификации */
  private readonly tokenFactory: () => Promise<string>;

  /** Формат URL для миниатюр пользовательских аватарок */
  private readonly thumbnailUrlFormat: string;

  /**
   * Конструктор класса RoomHub
   * @param url - URL endpoint SignalR хаба
   * @param tokenFactory - Функция для получения токена аутентификации
   * @param thumbnailUrlFormat - Формат URL для миниатюр пользователей
   */
  constructor(url: string, tokenFactory: () => Promise<string>, thumbnailUrlFormat: string) {
    this.handlers = [];
    this.url = url;
    this.tokenFactory = tokenFactory;
    this.thumbnailUrlFormat = thumbnailUrlFormat;
  }

  /**
   * Подключается к SignalR хабу комнаты
   * @returns Promise, который разрешается после успешного подключения
   */
  async start(): Promise<void> {
    // Создаем и настраиваем подключение к хабу
    this.connection = new signalR.HubConnectionBuilder()
      .withUrl(this.url, {
        accessTokenFactory: this.tokenFactory,
      })
      .configureLogging(signalR.LogLevel.Information)
      .build();

    // Подписываемся на поток событий от сервера
    this.connection.on('Event', this.pushEvent.bind(this));

    // Запускаем подключение
    await this.connection.start();
  }

  /**
   * Обрабатывает входящие события и распространяет их по обработчикам
   * @param ev - Контейнер с данными события
   */
  private pushEvent(ev: RoomEventContainer): void {
    // Обрабатываем событие подключения - формируем URL аватарки
    if (ev.joinEvent) {
      if (ev.joinEvent.viewer.photoKey) {
        ev.joinEvent.viewer.photoUrl = this.thumbnailUrlFormat + ev.joinEvent.viewer.photoKey;
      }
    }
    // Обрабатываем событие комнаты - формируем URL аватарок всех пользователей
    else if (ev.roomEvent) {
      ev.roomEvent.room.viewers.forEach((v) => {
        if (v.photoKey) v.photoUrl = this.thumbnailUrlFormat + v.photoKey;
      });
    }

    // Передаем событие всем зарегистрированным обработчикам
    this.handlers.forEach((h) => h(ev));
  }

  /**
   * Добавляет обработчик событий комнаты
   * @param handler - Функция-обработчик событий
   */
  addHandler(handler: (event: RoomEventContainer) => void): void {
    this.handlers.push(handler);
  }

  /**
   * Удаляет обработчик событий комнаты
   * @param handler - Функция-обработчик для удаления
   */
  removeHandler(handler: (event: RoomEventContainer) => void): void {
    const index = this.handlers.indexOf(handler);
    if (index >= 0) this.handlers.splice(index, 1);
  }

  // Методы для отправки команд на сервер:

  /**
   * Подключается к комнате по идентификатору.
   * Отправляет сигнал на сервер о намерении подключиться к указанной комнате.
   * @param roomId - Идентификатор комнаты для подключения
   * @returns Promise, который разрешается после отправки запроса на сервер
   */
  async connect(roomId: string): Promise<void> {
    await this.connection?.send('Connect', roomId);
  }

  /**
   * Запрашивает данные текущей комнаты с сервера.
   * Используется для синхронизации состояния комнаты после подключения.
   * @returns Promise, который разрешается после отправки запроса на сервер
   */
  async getRoom(): Promise<void> {
    await this.connection?.send('GetRoom');
  }

  /**
   * Получает список сообщений чата комнаты.
   * Если указаны параметры, выборка начинается с конкретного сообщения и ограничивается указанным количеством.
   * @param fromId - Идентификатор сообщения, с которого начинать выборку (опционально)
   * @param count - Количество сообщений для получения (опционально)
   * @returns Promise, который разрешается после отправки запроса на сервер
   */
  async getMessages(fromId?: string, count?: number): Promise<void> {
    await this.connection?.send('GetMessages', fromId, count);
  }

  /**
   * Устанавливает текущий эпизод для просмотра
   * @param season - Номер сезона
   * @param series - Номер серии
   * @returns Promise, который разрешается после отправки команды
   */
  async setEpisode(season: number, series: number): Promise<void> {
    await this.connection?.send('SetEpisode', season, series);
  }

  /**
   * Отправляет текстовое сообщение в чат комнаты
   * @param message - Текст сообщения
   * @returns Promise, который разрешается после отправки сообщения
   */
  async sendMessage(message: string): Promise<void> {
    await this.connection?.send('SendMessage', message);
  }

  /**
   * Устанавливает позицию воспроизведения видео
   * @param ticks - Позиция в тиках
   * @returns Promise, который разрешается после отправки команды
   */
  async setTimeLine(ticks: number): Promise<void> {
    await this.connection?.send('SetTimeLine', ticks);
  }

  /**
   * Устанавливает состояние паузы/воспроизведения
   * @param pause - Флаг паузы (true - пауза, false - воспроизведение)
   * @param ticks - Текущая позиция в тиках
   * @param buffering - Флаг буферизации
   * @returns Promise, который разрешается после отправки команды
   */
  async setPause(pause: boolean, ticks: number, buffering: boolean): Promise<void> {
    await this.connection?.send('SetPause', pause, ticks, buffering);
  }

  /**
   * Устанавливает полноэкранный режим
   * @param fullScreen - Флаг полноэкранного режима
   * @returns Promise, который разрешается после отправки команды
   */
  async setFullScreen(fullScreen: boolean): Promise<void> {
    await this.connection?.send('SetFullScreen', fullScreen);
  }

  /**
   * Устанавливает скорость воспроизведения видео
   * @param speed - Скорость воспроизведения
   * @returns Promise, который разрешается после отправки команды
   */
  async setSpeed(speed: number): Promise<void> {
    await this.connection?.send('SetSpeed', speed);
  }

  /**
   * Устанавливает уровень громкости (заглушает/включает звук)
   * @param muted - Флаг заглушения звука (true - звук заглушен, false - звук включен)
   * @returns Promise, который разрешается после отправки команды
   */
  async setMuted(muted: boolean): Promise<void> {
    await this.connection?.send('SetMuted', muted);
  }

  /**
   * Отправляет звуковой сигнал "beep" указанному пользователю
   * @param target - Идентификатор целевого пользователя
   * @returns Promise, который разрешается после отправки сигнала
   */
  async beep(target: string): Promise<void> {
    await this.connection?.send('Beep', target);
  }

  /**
   * Отправляет звуковой сигнал "scream" указанному пользователю
   * @param target - Идентификатор целевого пользователя
   * @returns Promise, который разрешается после отправки сигнала
   */
  async scream(target: string): Promise<void> {
    await this.connection?.send('Scream', target);
  }

  /**
   * Уведомляет сервер о наборе текста пользователем
   * @returns Promise, который разрешается после отправки уведомления
   */
  async type(): Promise<void> {
    await this.connection?.send('Type');
  }

  /**
   * Синхронизирует состояние медиа-контента при подключении пользователя к комнате
   * @returns Promise, который разрешается после отправки запроса синхронизации
   */
  async sync(): Promise<void> {
    await this.connection?.send('Sync');
  }

  /**
   * Отключается от SignalR хаба комнаты
   * @returns Promise, который разрешается после отключения
   */
  async disconnect(): Promise<void> {
    await this.connection?.stop();
  }
}

File: Rooms.Infrastructure.Web/Rooms/Hubs/RoomHub.cs
using System.Security.Claims;
using MediatR;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.SignalR;
using Rooms.Application.Abstractions.Commands;
using Rooms.Application.Abstractions.Exceptions;
using Rooms.Application.Abstractions.Queries;
using Rooms.Application.Abstractions.RoomEvents.Messages;
using Rooms.Application.Abstractions.RoomEvents.Room;
using Rooms.Domain.Rooms.Exceptions;
using Rooms.Infrastructure.Web.Metrics;
using Rooms.Infrastructure.Web.Rooms.Exceptions;

namespace Rooms.Infrastructure.Web.Rooms.Hubs;

/// <summary>
/// SignalR хаб для управления комнатами и взаимодействия между пользователями
/// </summary>
/// <param name="mediator">Экземпляр медиатора для обработки CQRS запросов</param>
[Authorize]
public class RoomHub(ISender mediator) : Hub
{
    /// <summary>
    /// Подключение пользователя к комнате
    /// </summary>
    /// <param name="roomId">Идентификатор комнаты</param>
    public async Task Connect(Guid roomId)
    {
        // Получаем идентификатор текущего пользователя
        var userId = GetUserId();

        // Устанавливаем статус "онлайн" для пользователя в комнате
        await mediator.Send(new SetOnlineCommand
        {
            RoomId = roomId,
            Online = true,
            ViewerId = userId
        });

        // Сохраняем идентификатор комнаты в контексте подключения
        Context.Items.Add("roomId", roomId);

        // Добавляем подключение к группе SignalR для комнаты
        await Groups.AddToGroupAsync(Context.ConnectionId, roomId.ToString());

        // Отправляем информацию о подключении вызывающему клиенту
        await Clients.Caller.SendAsync("Event", new ConnectEvent());
        
        // Увеличение метрики
        RoomsConnectionMetrics.Increment();
    }

    /// <summary>
    /// Получить данные комнаты
    /// </summary>
    public async Task GetRoom()
    {
        // Получаем идентификаторы пользователя и комнаты из контекста подключения
        var userId = GetUserId();
        var roomId = GetRoomId();
        
        // Получаем информацию о комнате
        var room = await mediator.Send(new GetRoomByIdQuery { RoomId = roomId, ViewerId = userId });

        // Создаем событие
        var @event = new RoomEvent { Room = room };

        // Отправляем информацию о комнате вызывающему клиенту
        await Clients.Caller.SendAsync("Event", @event);
    }
    
    /// <summary>
    /// Подключение пользователя к комнате и синхронизация текущего состояния медиа-контента
    /// </summary>
    public async Task Sync()
    {
        // Получаем идентификаторы пользователя и комнаты из контекста подключения
        var userId = GetUserId();
        var roomId = GetRoomId();

        // Запрашиваем текущее синхронизированное состояние комнаты
        var data = await mediator.Send(new GetRoomSyncDataQuery { Id = roomId, ViewerId = userId });

        // Последовательно отправляем все события состояния для синхронизации клиента

        // Отправляем событие текущего эпизода (только для сериалов)
        if (data.EpisodeEvent != null)
            await Clients.Caller.SendAsync("Event", data.EpisodeEvent);
        
        // Отправляем событие состояния паузы/воспроизведения
        await Clients.Caller.SendAsync("Event", data.PauseEvent);

        // Отправляем событие текущей позиции воспроизведения (таймкод)
        await Clients.Caller.SendAsync("Event", data.TimeLineEvent);

        // Отправляем событие скорости воспроизведения
        await Clients.Caller.SendAsync("Event", data.SpeedEvent);
    }

    /// <summary>
    /// Получение сообщений комнаты
    /// </summary>
    /// <param name="fromMessageId">Идентификатор сообщения, начиная с которого нужно получить сообщения (опционально)</param>
    /// <param name="count">Количество сообщений для получения (опционально, по умолчанию 20, максимум 50)</param>
    public async Task GetMessages(Guid? fromMessageId, int? count)
    {
        // Устанавливаем значения по умолчанию для количества сообщений
        if (count is null or < 0) count = 20;
        else if (count > 50) count = 50;

        // Получаем идентификаторы пользователя и комнаты
        var userId = GetUserId();
        var roomId = GetRoomId();

        // Получаем сообщения из базы данных
        var messages = await mediator.Send(new GetRoomMessagesQuery
        {
            RoomId = roomId,
            FromMessageId = fromMessageId,
            Count = count.Value,
            ViewerId = userId
        });

        // Создаем событие
        var @event = new MessagesEvent { Messages = messages };

        // Отправляем сообщения вызывающему клиенту
        await Clients.Caller.SendAsync("Event", @event);
    }

    /// <summary>
    /// Изменение текущей серии в комнате
    /// </summary>
    /// <param name="season">Номер сезона</param>
    /// <param name="episode">Номер серии</param>
    public async Task SetEpisode(int season, int episode)
    {
        // Получаем идентификаторы пользователя и комнаты
        var userId = GetUserId();
        var roomId = GetRoomId();

        // Отправляем команду на изменение серии
        await mediator.Send(new SetEpisodeCommand
        {
            Season = season,
            Episode = episode,
            ViewerId = userId,
            RoomId = roomId
        });
    }

    /// <summary>
    /// Уведомление о наборе сообщения пользователем
    /// </summary>
    public async Task Type()
    {
        // Получаем идентификаторы пользователя и комнаты
        var userId = GetUserId();
        var roomId = GetRoomId();

        // Отправляем команду о наборе сообщения
        await mediator.Send(new TypingCommand
        {
            ViewerId = userId,
            RoomId = roomId
        });
    }

    /// <summary>
    /// Отправка сообщения в комнату
    /// </summary>
    /// <param name="text">Текст сообщения</param>
    public async Task SendMessage(string text)
    {
        // Получаем идентификаторы пользователя и комнаты
        var userId = GetUserId();
        var roomId = GetRoomId();

        // Отправляем сообщение через медиатор
        await mediator.Send(new SendMessageCommand
        {
            Message = text,
            ViewerId = userId,
            RoomId = roomId,
            ConnectionId = Context.ConnectionId
        });
    }

    /// <summary>
    /// Установка текущей позиции воспроизведения
    /// </summary>
    /// <param name="ticks">Позиция</param>
    public async Task SetTimeLine(long ticks)
    {
        // Получаем идентификаторы пользователя и комнаты
        var userId = GetUserId();
        var roomId = GetRoomId();

        // Отправляем команду на установку позиции воспроизведения
        await mediator.Send(new SetTimeLineCommand
        {
            TimeLine = TimeSpan.FromTicks(ticks),
            ViewerId = userId,
            RoomId = roomId
        });
    }

    /// <summary>
    /// Установка состояния паузы
    /// </summary>
    /// <param name="pause">Флаг паузы (true - поставить на паузу)</param>
    /// <param name="ticks">Текущая позиция воспроизведения</param>
    /// <param name="buffering">Флаг, что пауза вызвана дозагрузкой контента</param>
    public async Task SetPause(bool pause, long ticks, bool buffering)
    {
        // Получаем идентификаторы пользователя и комнаты
        var userId = GetUserId();
        var roomId = GetRoomId();

        // Отправляем команду на установку паузы
        await mediator.Send(new SetPauseCommand
        {
            TimeLine = TimeSpan.FromTicks(ticks),
            Pause = pause,
            ViewerId = userId,
            RoomId = roomId,
            Buffering = buffering
        });
    }

    /// <summary>
    /// Установка скорости воспроизведения
    /// </summary>
    /// <param name="speed">Скорость воспроизведения (1.0 - нормальная скорость)</param>
    public async Task SetSpeed(double speed)
    {
        // Получаем идентификаторы пользователя и комнаты
        var userId = GetUserId();
        var roomId = GetRoomId();

        // Отправляем команду на установку скорости воспроизведения
        await mediator.Send(new SetSpeedCommand
        {
            ViewerId = userId,
            RoomId = roomId,
            Speed = speed
        });
    }

    /// <summary>
    /// Установка уровня громкости
    /// </summary>
    /// <param name="muted">Уровень громкости (0-100)</param>
    public async Task SetMuted(bool muted)
    {
        // Получаем идентификаторы пользователя и комнаты
        var userId = GetUserId();
        var roomId = GetRoomId();

        // Отправляем команду на установку громкости
        await mediator.Send(new SetVolumeCommand
        {
            ViewerId = userId,
            RoomId = roomId,
            Muted = muted
        });
    }

    /// <summary>
    /// Установка полноэкранного режима
    /// </summary>
    /// <param name="fullScreen">Флаг полноэкранного режима</param>
    public async Task SetFullScreen(bool fullScreen)
    {
        // Получаем идентификаторы пользователя и комнаты
        var userId = GetUserId();
        var roomId = GetRoomId();

        // Отправляем команду на установку полноэкранного режима
        await mediator.Send(new SetFullscreenCommand
        {
            Fullscreen = fullScreen,
            ViewerId = userId,
            RoomId = roomId
        });
    }

    /// <summary>
    /// Отправка звукового сигнала (бип) другому пользователю
    /// </summary>
    /// <param name="target">Идентификатор целевого пользователя</param>
    public async Task Beep(Guid target)
    {
        // Проверяем кулдаун на действия
        Action();

        // Получаем идентификаторы пользователя и комнаты
        var userId = GetUserId();
        var roomId = GetRoomId();

        // Отправляем команду на отправку бипа
        await mediator.Send(new BeepCommand
        {
            ViewerId = userId,
            RoomId = roomId,
            TargetId = target
        });
    }

    /// <summary>
    /// Отправка сигнала "крик" другому пользователю
    /// </summary>
    /// <param name="target">Идентификатор целевого пользователя</param>
    public async Task Scream(Guid target)
    {
        // Проверяем кулдаун на действия
        Action();

        // Получаем идентификаторы пользователя и комнаты
        var userId = GetUserId();
        var roomId = GetRoomId();

        // Отправляем команду на отправку крика
        await mediator.Send(new ScreamCommand
        {
            ViewerId = userId,
            RoomId = roomId,
            TargetId = target
        });
    }

    /// <summary>
    /// Обработчик отключения пользователя
    /// </summary>
    /// <param name="exception">Исключение, если отключение было вызвано ошибкой</param>
    public override async Task OnDisconnectedAsync(Exception? exception)
    {
        // Получаем идентификаторы пользователя и комнаты
        var userId = GetUserId();
        var roomId = GetRoomId();

        try
        {
            // Устанавливаем статус "оффлайн" для пользователя
            await mediator.Send(new SetOnlineCommand
            {
                Online = false,
                RoomId = roomId,
                ViewerId = userId
            });
        }
        catch (RoomNotFoundException)
        {
            // Игнорируем, так как комната может быть удалена
        }
        catch (ViewerNotFoundException)
        {
            // Игнорируем, так как пользователь может быть исключен / сам вышел
        }

        // Вызываем базовую реализацию метода
        await base.OnDisconnectedAsync(exception);
        
        // Уменьшение метрики
        RoomsConnectionMetrics.Decrement();
    }

    /// <summary>
    /// Получение идентификатора текущего пользователя из контекста
    /// </summary>
    /// <returns>Идентификатор пользователя</returns>
    /// <exception cref="InvalidOperationException">Если идентификатор не найден</exception>
    private Guid GetUserId()
    {
        var id = Context.User?.FindFirstValue(ClaimTypes.NameIdentifier) ?? throw new InvalidOperationException();
        return Guid.Parse(id);
    }

    /// <summary>
    /// Получение идентификатора комнаты из контекста подключения
    /// </summary>
    /// <returns>Идентификатор комнаты</returns>
    /// <exception cref="InvalidOperationException">Если идентификатор комнаты не найден</exception>
    private Guid GetRoomId()
    {
        return (Guid)(Context.Items["roomId"] ?? throw new InvalidOperationException());
    }

    /// <summary>
    /// Проверка кулдауна между действиями пользователя
    /// </summary>
    /// <exception cref="ActionCooldownException">Если действие выполняется слишком часто</exception>
    private void Action()
    {
        // Получаем время последнего действия
        var date = (DateTime?)Context.Items["LastActionTime"] ?? DateTime.MinValue;

        // Получаем текущее время
        var now = DateTime.Now;

        // Вычисляем разницу между текущим временем и временем последнего действия
        var difference = now - date;

        // Проверяем, что с последнего действия прошла хотя бы минута
        if (difference.TotalSeconds < 30) throw new ActionCooldownException(30 - difference.Seconds);

        // Обновляем время последнего действия
        Context.Items["LastActionTime"] = now;
    }
}